src/Reflection/ParametersAcceptorSelector.php

```diff
@@ -126,33 +127,51 @@ public static function selectFromArgs(
 
 			if (count($args) >= 3 && (bool) $args[0]->getAttribute(CurlSetOptArgVisitor::ATTRIBUTE_NAME)) {
 				$optType = $scope->getType($args[1]->value);
-				if ($optType instanceof ConstantIntegerType) {
-					$optValueType = self::getCurlOptValueType($optType->getValue());
-
-					if ($optValueType !== null) {
-						$acceptor = $parametersAcceptors[0];
-						$parameters = $acceptor->getParameters();
-
-						$parameters[2] = new NativeParameterReflection(
-							$parameters[2]->getName(),
-							$parameters[2]->isOptional(),
-							$optValueType,
-							$parameters[2]->passedByReference(),
-							$parameters[2]->isVariadic(),
-							$parameters[2]->getDefaultValue(),
-						);
-
-						$parametersAcceptors = [
-							new FunctionVariant(
-								$acceptor->getTemplateTypeMap(),
-								$acceptor->getResolvedTemplateTypeMap(),
-								array_values($parameters),
-								$acceptor->isVariadic(),
-								$acceptor->getReturnType(),
-								$acceptor instanceof ExtendedParametersAcceptor ? $acceptor->getCallSiteVarianceMap() : TemplateTypeVarianceMap::createEmpty(),
-							),
-						];
+
+				$optValueType = null;
+				foreach ($optType->getConstantScalarValues() as $scalarValue) {
+					if (!is_int($scalarValue)) {
+						$optValueType = null;
+						break;
+					}
+
+					$valueType = self::getCurlOptValueType($scalarValue);
+					if ($valueType === null) {
+						$optValueType = null;
+						break;
+					}
+
+					if ($optValueType === null) {
+						$optValueType = $valueType;
+						continue;
 					}
+
+					$optValueType = TypeCombinator::union($optValueType, $valueType);
```

Comment:
It'd be more performant to call `::union` just once once we gather all the types to union.